### PR TITLE
fix dead lock in SetLoopBodyEntryPoint

### PR DESCRIPTION
**Pull request by leirocks**
should not aquire FunctionProxy global lock while out of collecting code gen data phase. adding assertion to enforce that

### CODE REVIEW

File: lib/Runtime/Base/FunctionBody.cpp [hunk 1]
```diff
@@ -8305,30 +8305,45 @@ namespace Js
...
-        if (!this->sharedPropertyGuards->TryGetValue(propertyId, &guard))
+        if (!localGuards->TryGetValue(propertyId, &guard))
         {
             ThreadContext* threadContext = scriptContext->GetThreadContext();
             guard = threadContext->RegisterSharedPropertyGuard(propertyId);
-            this->sharedPropertyGuards->Add(propertyId, guard);
+
+            struct AutoResizeLock
+            {
+                AutoResizeLock(SharedPropertyGuardDictionary* d) :d(d) { d->LockResize(); }
```
Comment (leirocks) [inline lib/Runtime/Base/FunctionBody.cpp]: you are correct. this is recycler allocated, should not worry the old buffer being deleted

File: lib/Runtime/Base/FunctionBody.cpp [hunk 2]
```diff
@@ -8336,14 +8329,6 @@ namespace Js
         if (sharedPropertyGuards != nullptr)
         {
             auto guards = AnewArray(alloc, Js::PropertyId, sharedPropertyGuards->Count());
-
-            struct AutoResizeLock
-            {
-                AutoResizeLock(SharedPropertyGuardDictionary* d) :d(d) { d->LockResize(); }
-                ~AutoResizeLock() { d->UnlockResize(); }
-                SharedPropertyGuardDictionary* d;
-            } autoResizeLock(sharedPropertyGuards);
```
Comment (MikeHolman) [inline lib/Runtime/Base/FunctionBody.cpp]: I think this one we still need resize lock

Comment (leirocks) [inline lib/Runtime/Base/FunctionBody.cpp]: just took a closer look, we might still need lock on both location. not for old buffer being deleted, but for the count and the iterator to be consistent. while adding to the dictionary, it increase the count first, and then write the buffer, if background thread finished a get operation in between, it'll get an count-1 entries

**Comment by MikeHolman** [comment]
:shipit:

**Pull request merged**